server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/projection/cassandra/EnqueuedMailsDao.java

```diff
@@ -0,0 +1,156 @@
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.mail.internet.AddressException;
+
+import org.apache.james.backends.cassandra.init.configuration.ClusterConfiguration;
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.backends.cassandra.utils.CassandraUtils;
+import org.apache.james.core.MailAddress;
+import org.apache.james.queue.rabbitmq.projection.cassandra.model.CassandraMailQueueItem;
+import org.apache.james.util.CompletableFutureUtil;
+import org.apache.mailet.PerRecipientHeaders;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.UDTValue;
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+
+public class EnqueuedMailsDao {
+
+    private final CassandraAsyncExecutor executor;
+    private final PreparedStatement selectFrom;
+    private final CassandraUtils cassandraUtils;
+    private final Integer bucketSize;
+
+    @Inject
+    public EnqueuedMailsDao(Session session, CassandraUtils cassandraUtils, ClusterConfiguration clusterConfiguration) {
+        this.executor = new CassandraAsyncExecutor(session);
+        this.cassandraUtils = cassandraUtils;
+        this.bucketSize = clusterConfiguration.getHosts().size();
+
+        this.selectFrom = prepareSelectFrom(session);
+    }
+
+    private PreparedStatement prepareSelectFrom(Session session) {
+        return session.prepare(select()
+                .from(TABLE_NAME)
+                .where(eq(QUEUE_NAME, bindMarker(QUEUE_NAME)))
+                .and(gte(TIME_RANGE_START, bindMarker(TIME_RANGE_START)))
+                .and(eq(BUCKET_ID, bindMarker(BUCKET_ID))));
+    }
+
+    public CompletableFuture<Stream<CassandraMailQueueItem>> findEnqueuedMailsFrom(String queueName, Long timeRangeStartFrom) {
+        return CompletableFutureUtil.allOf(IntStream
+                .range(0, bucketSize)
+                .mapToObj(bucketId -> findEnqueuedMailsFrom(queueName, timeRangeStartFrom, bucketId)))
+            .thenApply(stream -> stream.flatMap(Function.identity()));
+    }
+
+    private CompletableFuture<Stream<CassandraMailQueueItem>> findEnqueuedMailsFrom(String queueName, Long timeRangeStartFrom, Integer bucketId) {
+        return executor.execute(
+            selectFrom.bind()
+                .setString(QUEUE_NAME, queueName)
+                .setLong(TIME_RANGE_START, timeRangeStartFrom)
+                .setInt(BUCKET_ID, bucketId))
+            .thenApply(resultSet -> cassandraUtils.convertToStream(resultSet)
+                .map(this::toMailQueueItem));
+    }
+
+    private CassandraMailQueueItem toMailQueueItem(Row row) {
```

Comment:
I copied all fields, and some methods to extract field from resultSet -> from `CassandraMailRepositoryMailDAO`, could you help me to check whether which field we only needed.

Reply:
You did well.

I wonder if we can avoid that duplication... Maybe using some Mail -> Row -> Mail abstraction, or abstracting mail detail storage using a specific Dao...

Reply:
I still can not resolve it because, it's not fully duplicated, and although the process of row -> mail nearly the same, but it happens with rows from two different tables which has some same fields, to keep it isolate, i will keep the duplication of code.